Lesson restructure (#24)
modular/max-llm-book: book/src/step_09.md

# Encode and decode tokens

<div class="note">

Learn to convert between text and token IDs using tokenizers and MAX tensors.

</div>

> In this step, you'll implement utility functions to bridge the gap between text and the token IDs your model operates on. The `tokenize_text` function converts strings to MAX tensors, while `decode_tokens` converts tensors back to readable text.

Reviewer: ```suggestion
In this step, you'll implement utility functions to bridge the gap between text and the token IDs your model operates on. The `tokenize_text()` function converts an input string into a tensor of token IDs, while `decode_tokens()` converts token IDs into a string.
```

Language models don't work with raw text. Instead, they process token IDs - integers that represent pieces of text according to a vocabulary. GPT-2 uses Byte Pair Encoding (BPE), which breaks text into subword units. For example, "Hello world" becomes `[15496, 995]` - two tokens representing the words.

You'll use the HuggingFace tokenizer to handle the text-to-token conversion, then wrap it with functions that work with MAX tensors. This separation keeps tokenization (a preprocessing step) separate from model inference (tensor operations).

## Understanding tokenization

Tokenization converts text to a list of integers. The GPT-2 tokenizer uses a vocabulary of 50,257 tokens, where common words get single tokens and rare words split into subwords.

The HuggingFace tokenizer provides an `encode` method that takes text and returns a Python list of token IDs. For example:

```python

> tokenizer.encode("Hello world")  # Returns [15496, 995]

Reviewer: I think it looks better if you capture it:
```suggestion
tokens = tokenizer.encode("Hello world")  # Returns [15496, 995]
```
